File: bindings/js/test_rpc.js
"use strict";

// payloads will be packed/unpacked as msgpack
const msgpack = require("msgpackr");
const sleep = require("sleep-promise");

const busrt = require("./busrt/src/busrt.js");

// RPC notification handler
async function on_notification(ev) {
  console.log(ev.frame.sender, ev.get_payload());
}

// RPC call handler
async function on_call(ev) {
  let method = ev.method.toString();
  if (method == "test") {
    return msgpack.pack({ ok: true });
  } else if (method == "err") {
    throw new busrt.RpcError(-777, "test error");
  } else {
    throw new busrt.RpcError(busrt.RPC_ERROR_CODE_METHOD_NOT_FOUND);
  }
}

// frame handler (broadcasts and topics)
async function on_frame(frame) {
  console.log(frame.sender, frame.get_payload());
}

async function main() {
  // create and connect a new client
  let bus = new busrt.Client("js");
  await bus.connect("/tmp/busrt.sock");
  // init RPC layer
  let rpc = new busrt.Rpc(bus);
  // init RPC handlers if incoming event handling is required
  rpc.on_notification = on_notification;
  rpc.on_call = on_call;
  rpc.on_frame = on_frame;
  // send test notification
  await rpc.notify("target", new busrt.RpcNotification("hello"));
  let payload = { test: 123 };
  // call rpc test method, no response required
  await rpc.call0(
    "target",
    new busrt.RpcRequest("test", msgpack.pack(payload))
  );
  // call rpc test method and wait for the response
  try {
    let request = await rpc.call(
      "target",
      new busrt.RpcRequest("test", msgpack.pack(payload))
    );
    let result = await request.wait_completed();
    console.log(result.get_payload().toString());
  } catch (err) {
    console.log(err.code, err.message.toString());
  }
  // handle local RPC methods while connected
  while (rpc.is_connected()) {
    await sleep(1000);
  }
  await bus.disconnect();
}

main();

File: bindings/python/busrt/tools/test_rpc.py
import busrt

import threading
import time
import random

iters = 10_000
workers = 1

cnt = 0


def test(w, iters):
    payload = b'\x01' * 1024
    global cnt
    name = f'test{random.randint(0,1000_000)}-{w}'
    path = '/tmp/busrt.sock'
    # path = 'localhost:9924'
    bus = busrt.client.Client(path, name)
    bus.timeout = 5
    bus.buf_size = 1024 * 1024
    bus.connect()
    rpc = busrt.rpc.Rpc(bus)
    print(f'Connected to {path}')
    for i in range(iters):
        result = rpc.call('y', busrt.rpc.Request('test',
                                         payload)).wait_completed(timeout=5)
        cnt += 1


started = time.perf_counter()
for w in range(workers):
    threading.Thread(target=test, args=(w, int(iters / workers))).start()
while cnt < iters:
    print(cnt)
    time.sleep(0.1)
elapsed = time.perf_counter() - started
speed = round(iters / elapsed)
print(f'{speed} iters/s ({round(1_000_000/speed)} ms per iter)')
